### PR TITLE
feat: [PLG-1861]: aws marketplace feature flags listing

### PR DESCRIPTION
**Pull request by mtran7**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42839)

### CODE REVIEW

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import software.wings.beans.Account;
+import software.wings.beans.Account.AccountKeys;
+import software.wings.service.intfc.AccountService;
+
+import com.google.inject.Inject;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+
+@OwnedBy(FF) // mt-unsure
```
Comment (hitesharinga) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java:33]: This we might not need. Check with FF team what happens if account expires

File: 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java [hunk 2]
```diff
@@ -633,7 +666,10 @@ public boolean updateLicenseForProduct(
     } else if (marketPlaceConfig.getAwsMarketPlaceCeProductCode().equals(productCode)) {
       updateCeLicense(
           accountId, CeLicenseInfo.builder().expiryTime(expirationTime).licenseType(CeLicenseType.PAID).build());
-    } else {
+    } else if (marketPlaceConfig.getAwsMarketPlaceFfProductCode().equals(productCode)) {
+      updateFfLicense(
```
Comment (hitesharinga) [inline 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java:670]: check with ff team which license we have to enable.

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                           .acceptableExecutionTime(ofMinutes(5))
+                           .handler(this)
+                           .filterExpander(query -> query.field(AccountKeys.ffLicenseInfo).exists())
+                           .schedulingType(REGULAR)
+                           .persistenceProvider(persistenceProvider)
+                           .redistribute(true));
+  }
+
+  @Override
+  public void registerIterator(IteratorExecutionHandler iteratorExecutionHandler) {
```
Comment (mtran7) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java:59]: Not needed

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseInfo.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import lombok.Value;
+
+@Value
+@Builder
+@OwnedBy(FF)
+public class FfLicenseInfo {
+  @JsonIgnore public static final int FF_TRIAL_GRACE_PERIOD_DAYS = 15;
+
+  private FfLicenseType licenseType;
+  private long expiryTime;
```
Comment (hitesharinga) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseInfo.java:26]: We don't need.

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (mtran7) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java:1]: Don't need entire file

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseType.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    public static final int FF_TRIAL_PERIOD_DAYS = 15;
+  }
+
+  FfLicenseType(int period, ChronoUnit periodUnit) {
+    this.defaultPeriod = period;
+    this.defaultPeriodUnit = periodUnit;
+  }
+
+  public long getDefaultExpiryTime() {
+    Calendar calendar = Calendar.getInstance();
```
Comment (hitesharinga) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseType.java:39]: don't need this

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseInfo.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2021 Harness Inc. All rights reserved.
```
Comment (mtran7) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseInfo.java:2]: Don't need this file

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseType.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (mtran7) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseType.java:1]: Don't need this file

File: 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java [hunk 9]
```diff
@@ -465,6 +467,37 @@ public boolean updateCeLicense(@NotEmpty String accountId, CeLicenseInfo ceLicen
...
+    }
+
+    FfLicenseInfo ffLicenseInfo = FfLicenseInfo.builder()
+                                      .licenseType(FfLicenseType.LIMITED_TRIAL)
+                                      .expiryTime(Math.max(FfLicenseType.LIMITED_TRIAL.getDefaultExpiryTime(),
+                                          FfLicenseType.getEndOfYearAsMillis(2020)))
+                                      .build();
+    updateFfLicense(accountId, ffLicenseInfo);
+    return true;
+  }
```
Comment (hitesharinga) [inline 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java:488]: dont need this

File: 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java (hunk 2 above)
Comment (mtran7) [inline 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java:670]: this is the main function that needs to be called when user tries to create an account

File: 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java [hunk 10]
```diff
@@ -465,6 +467,37 @@ public boolean updateCeLicense(@NotEmpty String accountId, CeLicenseInfo ceLicen
...
+
+  @Override
+  public boolean updateFfLicense(@NotEmpty String accountId, FfLicenseInfo ffLicenseInfo) {
+    accountDao.updateFfLicense(accountId, ffLicenseInfo);
+    if (Instant.now().toEpochMilli() < ffLicenseInfo.getExpiryTime()) {
```
Comment (mtran7) [inline 400-rest/src/main/java/software/wings/licensing/LicenseServiceImpl.java:493]: Replace w/ different logic that we'll get from PLG team

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  @Override
+  public void handle(Account account) {
+    FfLicenseInfo ffLicenseInfo =
+        Optional.ofNullable(account.getCeLicenseInfo()).orElse(FfLicenseInfo.builder().build());
+    long actualExpiryTime = ffLicenseInfo.getExpiryTime();
+    long expiryTime = ffLicenseInfo.getExpiryTimeWithGracePeriod();
+    if (expiryTime != 0L && Instant.now().toEpochMilli() > expiryTime) {
+      account.setCloudCostEnabled(false);
```
Comment (mtran7) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java:73]: should be ff specific handling

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      accountService.update(account);
+    }
+
+    if (actualExpiryTime == Long.MAX_VALUE && !account.isCloudCostEnabled()) {
+      account.setCloudCostEnabled(true);
+      accountService.update(account);
```
Comment (mtran7) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java:79]: should be ff specific handling

**Comment by mtran7** [comment]
trigger copyrightcheck

File: 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
+/*
```
Comment (mtran7) [inline 400-rest/src/main/java/io/harness/ff/license/FfLicenseExpiryHandler.java:1]: Remove this file

**Comment by mtran7** [comment]
trigger messagecheck

**Review by neelam-harness — CHANGES_REQUESTED** [review]
let's revert any changes in the config or yaml files that were due to `clang-format` 
otherwise the pr LGTM

**Review by neelam-harness — APPROVED** [review]
Thanks for making the nit changes as we discussed. This makes it more clean.
LGTM.

**Comment by mtran7** [comment]
trigger checkstylecodeformat

**Review by hitesharinga — DISMISSED** [review]

File: 270-delegate-service-app/src/test/resources/test-config-delegate-service.yml [hunk 14]
```diff
@@ -50,6 +50,7 @@ mktPlaceConfig:
   awsSecretKey: awsMktPlaceSecretKeyPlaceHolder
```
[the remark this answers is not in the document]
Comment (mtran7) [reply 270-delegate-service-app/src/test/resources/test-config-delegate-service.yml:50]: removing "awsMktPlaceFfProductCodePlaceHolder"

**Review by hitesharinga — APPROVED** [review]

**Review by neelam-harness — APPROVED** [review]

**Review by hitesharinga — APPROVED** [review]
> Reviewed 2 of 17 files at r1, 1 of 12 files at r15, 1 of 15 files at r17, 3 of 5 files at r20, 9 of 10 files at r21, 2 of 2 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42839)* status: 17 file(s) to review.

**Comment by mtran7** [comment]
Needed these changes as well 
https://github.com/harness/harness-core/pull/43965/files

**Pull request merged**